chain/network/src/snapshot_hosts/mod.rs

```diff
@@ -57,14 +182,59 @@ impl Inner {
         self.hosts.push(d.peer_id.clone(), d.clone());
         Some(d)
     }
+
+    // Try to insert up to max_entries_added more peers into the state part selector for this part ID
+    fn insert_part_hosts(
+        &mut self,
+        sync_hash: &CryptoHash,
+        shard_id: ShardId,
+        part_id: &PartId,
+        max_entries_added: usize,
+    ) {
+        let selector = self.state_part_selectors.get(&shard_id).unwrap();
+        let seen_peers = selector.seen_peers(part_id);
+
+        let mut new_peers = BTreeSet::new();
+        for (peer_id, info) in self.hosts.iter() {
+            if seen_peers.contains(peer_id)
+                || info.sync_hash != *sync_hash
+                || !info.shards.contains(&shard_id)
+            {
+                continue;
+            }
+            let score = priority_score(peer_id, part_id);
+            if new_peers.len() < max_entries_added {
+                new_peers.insert(PartPriority {
+                    peer_id: peer_id.clone(),
+                    score,
+                    times_returned: 0,
+                });
+            } else {
+                let p = PartPriority { peer_id: peer_id.clone(), score, times_returned: 0 };
+                let worst = new_peers.first().unwrap().clone();
+                if p > worst {
```

Comment:
It would be good to add some documentation and/or testing of the fact that this function inserts the yet-to-be-seen hosts **with the best priorities** (and not just some number of arbitrary yet-to-be-seen hosts), since this behavior is important for ensuring rough global consensus on which hosts should serve which parts.

We could also have some kind of test in which two instances of `PartPeerSelector` have very similar (but not exactly same) host lists. We would expect that the hosts known to both instances appear in the same order as a subsequence of their individual request orderings.

Reply:
so i think this is kinda tested at least somewhat because if you make this change:

```diff
diff --git a/chain/network/src/snapshot_hosts/mod.rs b/chain/network/src/snapshot_hosts/mod.rs
index dfc4a3410..3b183aa10 100644
--- a/chain/network/src/snapshot_hosts/mod.rs
+++ b/chain/network/src/snapshot_hosts/mod.rs
@@ -212,7 +212,7 @@ impl Inner {
             } else {
                 let p = PartPriority { peer_id: peer_id.clone(), score, times_returned: 0 };
                 let worst = new_peers.first().unwrap().clone();
-                if p > worst {
+                if p < worst {
                     new_peers.insert(p);
                     assert!(new_peers.remove(&worst));
                 }
```

then the test fails. But yeah maybe it would be good to add some more in the comment about this, and to add another test case with `part_selection_cache_batch_size=2` so we can see this more explicitly, so just added another commit, PTAL